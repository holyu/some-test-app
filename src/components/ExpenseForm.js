import React from "react";
import DatePicker from "react-date-picker/dist/entry.nostyle";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense
        ? (props.expense.amount / 100).toString().replace(".", ",")
        : "",
      createdAt: props.expense ? new Date(props.expense.createdAt) : new Date(),
      error: undefined
    };
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = e => {
    let amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\,\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide description and amount."
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount.replace(",", "."), 10) * 100,
        createdAt: this.state.createdAt.getTime(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Description"
          autoFocus
          className="text-input"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <DatePicker
          className="datepicker no-clear"
          onChange={this.onDateChange}
          value={this.state.createdAt}
          returnValue="start"
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          className="textarea"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          {this.props.expense ? (
            <button className="button">Save Expense</button>
          ) : (
            <button className="button">Add Expense</button>
          )}
        </div>
      </form>
    );
  }
}
