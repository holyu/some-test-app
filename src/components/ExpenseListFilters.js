import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";
import DatePicker from "react-date-picker/dist/entry.nostyle";

export class ExpenseListFilters extends React.Component {
  onStartDatesChange = startDate => {
    this.props.setStartDate(startDate);
  };
  onEndDatesChange = endDate => {
    this.props.setEndDate(endDate);
  };
  onTextChange = e => this.props.setTextFilter(e.target.value);
  onSortChange = e => {
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    } else if (e.target.value === "date") {
      this.props.sortByDate();
    }
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DatePicker
          onChange={this.onStartDatesChange}
          value={this.props.filters.startDate}
          returnValue="start"
        />
        <DatePicker
          onChange={this.onEndDatesChange}
          value={this.props.filters.endDate}
          returnValue="end"
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
