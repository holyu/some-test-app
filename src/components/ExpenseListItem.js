import React from "react";
import { Link } from "react-router-dom";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__sub-title">
        {new Intl.DateTimeFormat("pl-PL", {
          year: "numeric",
          month: "short",
          day: "numeric"
        }).format(createdAt)}
      </span>
    </div>
    <h3 className="list-item__data">
      {new Intl.NumberFormat("pl-PL", {
        style: "currency",
        currency: "PLN"
      }).format(amount / 100)}
    </h3>
  </Link>
);

export default ExpenseListItem;
