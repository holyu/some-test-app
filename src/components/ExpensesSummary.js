import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from "../selectors/expenses-total";

export const ExpensesSummary = ({ expensesTotal, expensesCount }) => {
  const expensedText = expensesCount === 1 ? "expense" : "expenses";
  return (
    <div>
      Viewing {expensesCount} {expensedText} totalling{" "}
      {numeral(expensesTotal / 100).format("$0,0.00")}
    </div>
  );
};

const mapStateToProps = state => {
  const allVisibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expensesTotal: selectExpensesTotal(allVisibleExpenses),
    expensesCount: allVisibleExpenses.length
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
