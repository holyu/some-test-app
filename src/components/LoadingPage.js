import React from "react";
import loaderImg from "../images/loader.gif";

export const LoadingPage = () => (
  <div className="loader">
    <img className="loader__image" src={loaderImg} />
  </div>
);

export default LoadingPage;
