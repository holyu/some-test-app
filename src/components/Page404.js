import React from "react";
import { Link } from "react-router-dom";

export default () => (
  <div>
    This is 404! - <Link to="/">Home</Link>
  </div>
);
