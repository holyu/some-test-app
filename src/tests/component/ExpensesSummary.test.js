import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";

test("should render ExpensesSummary with 2 expenses", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesTotal={9436} expensesCount={2} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesSummary with 1 expenses", () => {
  const wrapper = shallow(
    <ExpensesSummary expensesTotal={9436} expensesCount={1} />
  );
  expect(wrapper).toMatchSnapshot();
});
