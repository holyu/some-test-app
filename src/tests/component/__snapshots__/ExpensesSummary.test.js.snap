// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expenses 1`] = `
<div>
  Viewing 
  1
   
  expense
   totalling
   
  $94.36
</div>
`;

exports[`should render ExpensesSummary with 2 expenses 1`] = `
<div>
  Viewing 
  2
   
  expenses
   totalling
   
  $94.36
</div>
`;
